### PR TITLE
Grouped Folders: Inbox

### PR DESCRIPTION
**Pull request by marbetschar**
This PR eventually fixes https://github.com/elementary/mail/issues/449.

Right now it only adds a unified Inbox, however we may want to iterate on this a bit more to expand support to other well known folders if possible (such as Drafts, Outbox, Sent, Archive, Junk and Trash).

Please note the current implementation uses a hardcoded folder named "INBOX". While this seems to work for this particular folder, this does not seem to be a reliable approach for other well known folders. There is already a [open PR](https://github.com/elementary/mail/pull/545) which provides an improved folder detection mechanism - so that's probably a good starting point.

Screenshot from 2021-05-04 07-56-03

### CODE REVIEW

**Comment by danrabbit** [comment]
I can confirm that this works as expected! Can we add this to a header like "All Mailboxes" so it's a bit more clear what this is?

**Comment by marbetschar** [comment]
@danrabbit added a "All Mailboxes" section header.

**Comment by hanaral** [comment]
Would it be possible to add an 'all sent' button as well? iOS Mail has the ability to add/remove quite a few different unified section and I think it would be a good starting point to add the main 2

**Comment by danrabbit** [comment]
@hanaral We usually prefer to have many smaller PRs than to try to do everything at once in a single branch. This makes code review faster and easier and helps with finding issues down the road if they occur

**Comment by marbetschar** [comment]
@danrabbit / @hanaral I digged into this a bit more and added support for `INBOX`, `Drafts`, `Sent` and `Archive` folders. Please note, showing a badge with unread messages is still not supported yet.

**Review by danrabbit — CHANGES_REQUESTED** [review]
Inbox works here, but I'm getting a number of errors for the other mailboxes. I think it would be best to go back to just inbox in this branch and then do follow up branches for other other mailbox types

on launch I'm getting these:
```
** (io.elementary.mail:3157): CRITICAL **: 11:31:11.615: mail_grouped_folder_source_item_build_folder_full_name: assertion 'folder_uri != NULL' failed

** (io.elementary.mail:3157): WARNING **: 11:31:11.615: GroupedFolderSourceItem.vala:134: Grouped folder '(null)' not available for account 'daniel@elementary.io'

** (io.elementary.mail:3157): CRITICAL **: 11:31:11.619: mail_grouped_folder_source_item_build_folder_full_name: assertion 'folder_uri != NULL' failed

** (io.elementary.mail:3157): WARNING **: 11:31:11.619: GroupedFolderSourceItem.vala:134: Grouped folder '(null)' not available for account 'daniel.p.fore@gmail.com'

(io.elementary.mail:3157): camel-WARNING **: 11:31:13.398: CamelIMAPXStore::get_folder_info_sync() reported failure without setting its GError

** (io.elementary.mail:3157): WARNING **: 11:31:13.398: GroupedFolderSourceItem.vala:134: Grouped folder '%5bGmail%5d/Drafts' not available for account 'daniel@elementary.io'

(io.elementary.mail:3157): camel-WARNING **: 11:31:14.585: CamelIMAPXStore::get_folder_info_sync() reported failure without setting its GError

** (io.elementary.mail:3157): WARNING **: 11:31:14.585: GroupedFolderSourceItem.vala:134: Grouped folder '%5bGmail%5d/Drafts' not available for account 'daniel.p.fore@gmail.com'
```

when clicking sent:
```
g_variant_new_string: assertion 'string != NULL' failed
```

**Comment by danrabbit** [comment]
Another small thing I'm noticing is that when we first open Mail, it restores the selected folder for example if you select Inbox under one of your accounts. But if the last selected folder was one of these grouped folders, it does not restore on startup

**Comment by marbetschar** [comment]
- Reverted to INBOX only
- Automatically select INBOX on app start
- Merged latest master

File: src/ConversationList/ConversationListBox.vala [hunk 1]
```diff
@@ -253,20 +271,33 @@ public class Mail.ConversationListBox : VirtualizingListBox {
     }
 
     public async int archive_selected_messages () {
-        var threads = new Gee.ArrayList<Camel.FolderThreadNode?> ();
+        var archive_threads = new Gee.HashMap<string, Gee.ArrayList<Camel.FolderThreadNode?>> ();
         var selected_rows = get_selected_rows ();
-        foreach (unowned GLib.Object row in selected_rows) {
-            threads.add (((ConversationItemModel)row).node);
+        foreach (var selected_row in selected_rows) {
```
Comment (danrabbit) [inline src/ConversationList/ConversationListBox.vala:276]: should this still be:
```suggestion
        foreach (unowned var selected_row in selected_rows) {
```

File: src/ConversationList/ConversationListBox.vala [hunk 2]
```diff
@@ -276,13 +307,22 @@ public class Mail.ConversationListBox : VirtualizingListBox {
     }
 
     public int trash_selected_messages () {
-        var threads = new Gee.ArrayList<Camel.FolderThreadNode?> ();
+        var trash_threads = new Gee.HashMap<string, Gee.ArrayList<Camel.FolderThreadNode?>> ();
+
         var selected_rows = get_selected_rows ();
-        foreach (var row in selected_rows) {
-            threads.add (((ConversationItemModel)row).node);
+        foreach (var selected_row in selected_rows) {
```
Comment (danrabbit) [inline src/ConversationList/ConversationListBox.vala:313]: unowned again probably?
```suggestion
        foreach (unowned var selected_row in selected_rows) {
```

**Review by danrabbit — COMMENTED** [review]
This works as expected for me, nice job! Only two very small code comments

**Comment by marbetschar** [comment]
@danrabbit made the fixes as suggested. Everything still works on my end, so I guess we are good. Still not 100% sure about when to use `unowned` though :)

**Review by danrabbit — APPROVED** [review]
LFG :rocket:

**Pull request merged**